### ISSUE TITLE
Use XDG Base Directory on Linux

### ISSUE THREAD

**Issue by CodingKoopa** [libgdx/libgdx#6559]
#### Issue details
Currently, by default, libGDX [Preferences](https://github.com/libgdx/libgdx/wiki/Preferences) are saved to `~/.prefs` on Linux, which is nonstandard. The [XDG Base Directory](https://wiki.archlinux.org/title/XDG_Base_Directory) specification suggests that, to determine where to write configuration files to, programs should:
- First try the environment variable `$XDG_CONFIG_HOME`.
- If the environment variable is unset, default to `$HOME/.config`.

This standard serves to make the home directory cleaner, and normalize the way different programs save their configurations.

If this is adopted in libGDX, it would likely be necessary to check the (then legacy) `~/.prefs` path in addition to the above two paths, as to not break existing configurations.

This is somewhat similar to #6367, but not the same as this pertains to preferences rather than local storage.

#### Please select the affected platforms
- [ ] Android
- [ ] iOS
- [ ] HTML/GWT
- [ ] Windows
- [X] Linux
- [ ] MacOS

Thanks!

**Comment by lyze237** [comment]
While at it, it'd also be cool to save prefs in %appdata% on windows and in ~/Library/Preferences on macos. As those are also more recommended to save program preferences in.)

**Comment by tommyettinger** [comment]
Yeah, preferences aren't usually needed by users as files, so having them in the APPDATA folder on Windows isn't a problem, and using the XDG folder on Linux would clean up the home directory nicely. I'll go with what @lyze237 says re: MacOS, because I don't use it.

**Comment by MrStahlfelge** [comment]
So, looks like we just need some desktop devs that make it happen?

**Comment by tommyettinger** [comment]
A few questions remain -- if the old preferences folder is present, we should probably use it, to avoid making saved data un-loadable. So I suppose this change only applies to new folders of preferences. We still need to check the old folder, but should it have precedence? Should the new folder be preferred, since it would have been written more recently if present?

**Comment by lyze237** [comment]
I just googled and found this stack exchange answer, so took it from there. Up to you which one to pick: https://apple.stackexchange.com/questions/28928/what-is-the-macos-equivalent-to-windows-appdata-folder

**Comment by terefang** [comment]
taking the discussion off discord, i have so far

```java

public class OsUtil
{
    private static String OS = System.getProperty("os.name").toLowerCase();

    static public boolean isAndroid = System.getProperty("java.runtime.name").contains("Android");
    static public boolean isMac = !isAndroid && OS.contains("mac");
    static public boolean isWindows = !isAndroid && OS.contains("windows");
    static public boolean isLinux = !isAndroid && OS.contains("linux");
    static public boolean isIos = !isAndroid && (!(isWindows || isLinux || isMac)) || OS.startsWith("ios");

    static public boolean isARM = System.getProperty("os.arch").startsWith("arm") || System.getProperty("os.arch").startsWith("aarch64");
    static public boolean is64Bit = System.getProperty("os.arch").contains("64") || System.getProperty("os.arch").startsWith("armv8");

    public static boolean isGwt = false;

    static {
        try {
            Class.forName("com.google.gwt.core.client.GWT");
            isGwt = true;
        }
        catch(Exception ignored) { /* IGNORED */ }

        boolean isMOEiOS = "iOS".equals(System.getProperty("moe.platform.name"));
        if (isMOEiOS || (!isAndroid && !isWindows && !isLinux && !isMac)) {
            isIos = true;
            isAndroid = false;
            isWindows = false;
            isLinux = false;
            isMac = false;
            is64Bit = false;
        }
   }

    public static String getConfigDirectory(String applicationName)
    {
        String CONFIG_HOME = null;

        if((CONFIG_HOME = System.getenv("XDG_CONFIG_HOME"))==null)
        {
            if(isLinux || isAndroid)
            {
                CONFIG_HOME = System.getProperty("user.home")+"/.config";
            }
            else if(isMac)
            {
                CONFIG_HOME = System.getProperty("user.home")+"/Library/Preferences";
            }
            else if(isIos)
            {
                CONFIG_HOME = System.getProperty("user.home")+"/Documents";
            }
            else if(isWindows)
            {
                if((CONFIG_HOME = System.getenv("APPDATA"))==null)
                {
                    CONFIG_HOME = System.getProperty("user.home")+"/Local Settings";
                }
            }
        }

        if(applicationName==null || CONFIG_HOME==null) return CONFIG_HOME;

        return CONFIG_HOME+"/"+applicationName;
    }

    public static String getDataDirectory(String applicationName)
    {
        String DATA_HOME = null;

        if((DATA_HOME = System.getenv("XDG_DATA_HOME"))==null)
        {
            if(isLinux || isAndroid)
            {
                DATA_HOME = System.getProperty("user.home")+"/.local/share";
            }
            else if(isMac)
            {
                DATA_HOME = System.getProperty("user.home")+"/Library/Application Support";
            }
            else if(isIos)
            {
                DATA_HOME = System.getProperty("user.home")+"/Documents";
            }
            else if(isWindows)
            {
                if((DATA_HOME = System.getenv("APPDATA"))==null)
                {
                    DATA_HOME = System.getProperty("user.home")+"/Local Settings/Application Data";
                }
            }
        }

        if(applicationName==null || DATA_HOME==null) return DATA_HOME;

        return DATA_HOME+"/"+applicationName;
    }

}

```

**Comment by tommyettinger** [comment]
I like the idea here, but it overlaps a lot with the existing UIUtils. Maybe UIUtils could just call this, to avoid duplication, since this seems a little more robust. `isMOEiOS` should be moved to be next to the other isSomeOS fields, otherwise it isn't accessible -- unless you don't mean to expose whether the app is running with MOE or RoboVM, which seems like an important distinction in some cases. I like having a check for running on GWT, and this one makes sense to me (as much as anything GWT-related can). Generally, this seems useful for whenever a PR is written.

**Comment by lyze237** [comment]
There's already checks for type of platform (`Gdx.app.getType()` and `UIUtils`) so in my opinion there's not really a reason to do that all again.

Additionally according to @crykn as this should be a change to desktop only, it should only be in the lwjgl3 backend in the Lwjgl3Preferences and not in core.

**Comment by tommyettinger** [comment]
@lyze237 @crykn If it's in the LWJGL3 backend, it can only be used from that backend, which requires a lot more work to actually use in the core part of an app. Also, Lyze you're saying two opposite things -- there are already checks for the platform in core, but these checks don't belong in core...?

**Comment by crykn** [comment]
I agree with @lyze237. The backend/platform can already be queried by calling `Gdx.app.getType()` and the OS can be determined via `UIUtils`, so there is no need for this class.

> @tommyettinger: If it's in the LWJGL3 backend, it can only be used from that backend, which requires a lot more work to actually use in the core part of an app.

Why would the core part of the app need to know where the preferences are saved on desktop? I think this is an implementation detail of `Lwjgl3Preferences` (as well as `Lwjgl2Preferences` and `HeadlessPreferences`) and doesn't belong in core. Also the Android and iOS stuff is unnecessary as those backends don't save their preferences in certain directories, but rather use some platform specific stuff.

> @tommyettinger: Also, Lyze you're saying two opposite things -- there are already checks for the platform in core, but these checks don't belong in core...?

I think what Lyze meant is that there are already checks in core (i.e., `Gdx.app.getType()`, `UIUtils`) and that is all well, but the additional methods (`getUserDataDirectory`, `getUserConfigDirectory`) don't belong in core.

**Comment by lyze237** [comment]
> I think what Lyze meant is that there are already checks in core (i.e., Gdx.app.getType(), UIUtils) and that is all well, but the additional methods (getUserDataDirectory, getUserConfigDirectory) don't belong in core.

Yeah exactly that, also those should be used instead of checking everything manually.

> Why would the core part of the app need to know where the preferences are saved on desktop? [...] I don't think System.getProperty("os.name") works on GWT.

Yeah same opinion on that.

**Comment by tommyettinger** [comment]
Good catch on the `System.getProperty()`, and yeah I guess if the Preferences location is only used internally it can be inside the platform-specific parts. There's already platform-specific classes for the different Preferences where this could fit.

**Comment by terefang** [comment]
i have seen people posting on discord that they use Preferences in their apps.

does this count as internal too ?
